autowrap fortran fails with expression containing Mod function twice

Minimal example:

```
from sympy import Matrix, Mod,, symbols
from sympy.utilities.autowrap import autowrap

x, y = symbols(['x', 'y'])
expr = Mod(x, 2)- Mod(y, 2)
f = autowrap(expr, args=[x, y])
```

Error:
```
CodeWrapError: Error while executing command: /home/hen/miniconda/envs/py37/bin/python -c import numpy.f2py as f2py2e;f2py2e.main() -c -m wrapper_module_5 wrapped_code_5.f90. Command output is:
running build
running config_cc
unifing config_cc, config, build_clib, build_ext, build commands --compiler options
running config_fc
unifing config_fc, config, build_clib, build_ext, build commands --fcompiler options
running build_src
build_src
building extension "wrapper_module_5" sources
f2py options: []
f2py:> /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c
creating /tmp/tmpz9l25tyj/src.linux-x86_64-3.7
Reading fortran codes...
	Reading file 'wrapped_code_5.f90' (format:free)
Post-processing...
	Block: wrapper_module_5
			Block: autofunc
Post-processing (stage 2)...
Building modules...
	Building module "wrapper_module_5"...
		Creating wrapper for Fortran function "autofunc"("autofunc")...
		Constructing wrapper function "autofunc"...
		  autofunc = autofunc(x,y)
	Wrote C/API module "wrapper_module_5" to file "/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c"
	Fortran 77 wrappers are saved to "/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5-f2pywrappers.f"
  adding '/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c' to sources.
  adding '/tmp/tmpz9l25tyj/src.linux-x86_64-3.7' to include_dirs.
copying /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.c -> /tmp/tmpz9l25tyj/src.linux-x86_64-3.7
copying /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/f2py/src/fortranobject.h -> /tmp/tmpz9l25tyj/src.linux-x86_64-3.7
  adding '/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5-f2pywrappers.f' to sources.
build_src: building npy-pkg config files
running build_ext
customize UnixCCompiler
customize UnixCCompiler using build_ext
get_default_fcompiler: matching types: '['gnu95', 'intel', 'lahey', 'pg', 'absoft', 'nag', 'vast', 'compaq', 'intele', 'intelem', 'gnu', 'g95', 'pathf95', 'nagfor']'
customize Gnu95FCompiler
Found executable /usr/bin/gfortran
customize Gnu95FCompiler
customize Gnu95FCompiler using build_ext
building 'wrapper_module_5' extension
compiling C sources
C compiler: gcc -pthread -B /home/hen/miniconda/envs/py37/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC

creating /tmp/tmpz9l25tyj/tmp
creating /tmp/tmpz9l25tyj/tmp/tmpz9l25tyj
creating /tmp/tmpz9l25tyj/tmp/tmpz9l25tyj/src.linux-x86_64-3.7
compile options: '-I/tmp/tmpz9l25tyj/src.linux-x86_64-3.7 -I/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include -I/home/hen/miniconda/envs/py37/include/python3.7m -c'
gcc: /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c
gcc: /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c
In file included from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1823,
                 from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,
                 from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.h:13,
                 from /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c:15:
/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^~~~~~~
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c: In function ‘PyInit_wrapper_module_5’:
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c:333:3: warning: this ‘for’ clause does not guard... [-Wmisleading-indentation]
   for(i=0;f2py_routine_defs[i].name!=NULL;i++)
   ^~~
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/wrapper_module_5module.c:336:5: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘for’
     {
     ^
In file included from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/ndarraytypes.h:1823,
                 from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,
                 from /home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.h:13,
                 from /tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c:2:
/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it by " \
  ^~~~~~~
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c: In function ‘format_def’:
/tmp/tmpz9l25tyj/src.linux-x86_64-3.7/fortranobject.c:138:18: warning: comparison of integer expressions of different signedness: ‘Py_ssize_t’ {aka ‘long int’} and ‘long unsigned int’ [-Wsign-compare]
         if (size < sizeof(notalloc)) {
                  ^
compiling Fortran sources
Fortran f77 compiler: /usr/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -fPIC -O3 -funroll-loops
Fortran f90 compiler: /usr/bin/gfortran -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops
Fortran fix compiler: /usr/bin/gfortran -Wall -g -ffixed-form -fno-second-underscore -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops
compile options: '-I/tmp/tmpz9l25tyj/src.linux-x86_64-3.7 -I/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include -I/home/hen/miniconda/envs/py37/include/python3.7m -c'
gfortran:f90: wrapped_code_5.f90
error: Command "/usr/bin/gfortran -Wall -g -fno-second-underscore -fPIC -O3 -funroll-loops -I/tmp/tmpz9l25tyj/src.linux-x86_64-3.7 -I/home/hen/miniconda/envs/py37/lib/python3.7/site-packages/numpy/core/include -I/home/hen/miniconda/envs/py37/include/python3.7m -c -c wrapped_code_5.f90 -o /tmp/tmpz9l25tyj/wrapped_code_5.o" failed with exit status 1
wrapped_code_5.f90:6:13:

 REAL*8 :: Mod
             1
Error: Symbol ‘mod’ at (1) already has basic type of REAL
wrapped_code_5.f90:6:13:

 REAL*8 :: Mod
             1
Error: Symbol ‘mod’ at (1) already has basic type of REAL
```

And here's the generated code.
```
!******************************************************************************
!*                       Code generated with sympy 1.2                        *
!*                                                                            *
!*              See http://www.sympy.org/ for more information.               *
!*                                                                            *
!*                      This file is part of 'autowrap'                       *
!******************************************************************************

REAL*8 function autofunc(x, y)
implicit none
REAL*8, intent(in) :: x
REAL*8, intent(in) :: y

REAL*8 :: Mod
REAL*8 :: Mod
autofunc = Mod(1.0*x, 2.0) - Mod(1.0*y, 2.0)

end function
```

Clearly the problem is that Mod gets imported twice. Hopefully this is an easy fix but I don't know where to start.

--- a/sympy/external/tests/test_autowrap.py
+++ b/sympy/external/tests/test_autowrap.py
@@ -143,13 +143,22 @@ def runtest_issue_10274(language, backend):
                 ]
 
 
-def test_issue_14230():
+def test_issue_15230():
     has_module('f2py')
     x, y = symbols('x, y')
-    expr = Mod(x, 2) - Mod(y, -2)
-    f = autowrap(expr, args=[x, y])
+
+    expr = Mod(x, 3.0) - Mod(y, -2.0)
+    f = autowrap(expr, args=[x, y], language='F95',
+                 tempdir='/home/moorepants/Desktop/forty')
     exp_res = float(expr.xreplace({x: 3.5, y: 2.7}).evalf())
     assert abs(f(3.5, 2.7) - exp_res) < 1e-14
+
+    # TODO : This fails with integers.
+    #expr = Mod(x, 3) - Mod(y, -2)
+    #f = autowrap(expr, args=[x, y], language='F95',
+                 #tempdir='/home/moorepants/Desktop/forty')
+    #exp_res = expr.xreplace({x: 3, y: 2}).evalf()
+    #assert f(3, 2) == exp_res
 
 #
 # tests of language-backend combinations

--- a/sympy/external/tests/test_autowrap.py
+++ b/sympy/external/tests/test_autowrap.py
@@ -146,7 +146,7 @@ def runtest_issue_10274(language, backend):
 def test_issue_14230():
     has_module('f2py')
     x, y = symbols('x, y')
-    expr = Mod(x, 2) - Mod(y, 2)
+    expr = Mod(x, 2) - Mod(y, -2)
     f = autowrap(expr, args=[x, y])
     exp_res = float(expr.xreplace({x: 3.5, y: 2.7}).evalf())
     assert abs(f(3.5, 2.7) - exp_res) < 1e-14

--- a/sympy/external/tests/test_autowrap.py
+++ b/sympy/external/tests/test_autowrap.py
@@ -145,20 +145,18 @@ def runtest_issue_10274(language, backend):
 
 def test_issue_15230():
     has_module('f2py')
-    x, y = symbols('x, y')
 
+    x, y = symbols('x, y')
     expr = Mod(x, 3.0) - Mod(y, -2.0)
-    f = autowrap(expr, args=[x, y], language='F95',
-                 tempdir='/home/moorepants/Desktop/forty')
+    f = autowrap(expr, args=[x, y], language='F95')
     exp_res = float(expr.xreplace({x: 3.5, y: 2.7}).evalf())
     assert abs(f(3.5, 2.7) - exp_res) < 1e-14
 
-    # TODO : This fails with integers.
-    #expr = Mod(x, 3) - Mod(y, -2)
-    #f = autowrap(expr, args=[x, y], language='F95',
-                 #tempdir='/home/moorepants/Desktop/forty')
-    #exp_res = expr.xreplace({x: 3, y: 2}).evalf()
-    #assert f(3, 2) == exp_res
+    x, y = symbols('x, y', integer=True)
+    expr = Mod(x, 3) - Mod(y, -2)
+    f = autowrap(expr, args=[x, y], language='F95')
+    exp_res = expr.xreplace({x: 3, y: 2}).evalf()
+    assert f(3, 2) == exp_res
 
 #
 # tests of language-backend combinations

--- a/sympy/external/tests/test_autowrap.py
+++ b/sympy/external/tests/test_autowrap.py
@@ -155,8 +155,7 @@ def test_issue_15230():
     x, y = symbols('x, y', integer=True)
     expr = Mod(x, 3) - Mod(y, -2)
     f = autowrap(expr, args=[x, y], language='F95')
-    exp_res = expr.xreplace({x: 3, y: 2}).evalf()
-    assert f(3, 2) == exp_res
+    assert f(3, 2) == expr.xreplace({x: 3, y: 2})
 
 #
 # tests of language-backend combinations

--- a/sympy/external/tests/test_autowrap.py
+++ b/sympy/external/tests/test_autowrap.py
@@ -144,7 +144,7 @@ def runtest_issue_10274(language, backend):
 
 
 def test_issue_14230():
-
+    has_module('f2py')
     x, y = symbols('x, y')
     expr = Mod(x, 2) - Mod(y, 2)
     f = autowrap(expr, args=[x, y])

--- a/sympy/external/tests/test_autowrap.py
+++ b/sympy/external/tests/test_autowrap.py
@@ -2,7 +2,7 @@ import sympy
 import tempfile
 import os
 import warnings
-from sympy import symbols, Eq
+from sympy import symbols, Eq, Mod
 from sympy.external import import_module
 from sympy.tensor import IndexedBase, Idx
 from sympy.utilities.autowrap import autowrap, ufuncify, CodeWrapError
@@ -141,6 +141,15 @@ def runtest_issue_10274(language, backend):
                 "\n",
                 "}\n",
                 ]
+
+
+def test_issue_14230():
+
+    x, y = symbols('x, y')
+    expr = Mod(x, 2) - Mod(y, 2)
+    f = autowrap(expr, args=[x, y])
+    exp_res = float(expr.xreplace({x: 3.5, y: 2.7}).evalf())
+    assert abs(f(3.5, 2.7) - exp_res) < 1e-14
 
 #
 # tests of language-backend combinations

--- a/sympy/printing/fcode.py
+++ b/sympy/printing/fcode.py
@@ -59,7 +59,8 @@ known_functions = {
     "Abs": "abs",
     "conjugate": "conjg",
     "Max": "max",
-    "Min": "min"
+    "Min": "min",
+    "Mod": "mod",
 }
 
 

--- a/sympy/printing/fcode.py
+++ b/sympy/printing/fcode.py
@@ -312,7 +312,7 @@ class FCodePrinter(CodePrinter):
         if self._settings['standard'] in [66, 77]:
             msg = ("Python % operator and SymPy's Mod() function are not "
                    "supported by Fortran 66 or 77 standards.")
-            raise ValueError(msg)
+            raise NotImplementedError(msg)
         else:
             x, y = expr.args
             return "      modulo({}, {})".format(self._print(x), self._print(y))

--- a/sympy/printing/fcode.py
+++ b/sympy/printing/fcode.py
@@ -60,7 +60,6 @@ known_functions = {
     "conjugate": "conjg",
     "Max": "max",
     "Min": "min",
-    "Mod": "modulo",
 }
 
 
@@ -304,6 +303,19 @@ class FCodePrinter(CodePrinter):
             return self._print(eval_expr)
         else:
             return CodePrinter._print_Function(self, expr.func(*args))
+
+    def _print_Mod(self, expr):
+        # NOTE : Fortran has the functions mod() and modulo(). modulo() behaves
+        # the same wrt to the sign of the arguments as Python and SymPy's
+        # modulus computations (% and Mod()) but is not available in Fortran 66
+        # or Fortran 77, thus we raise an error.
+        if self._settings['standard'] in [66, 77]:
+            msg = ("Python % operator and SymPy's Mod() function are not "
+                   "supported by Fortran 66 or 77 standards.")
+            raise ValueError(msg)
+        else:
+            x, y = expr.args
+            return "      modulo({}, {})".format(self._print(x), self._print(y))
 
     def _print_ImaginaryUnit(self, expr):
         # purpose: print complex numbers nicely in Fortran.

--- a/sympy/printing/fcode.py
+++ b/sympy/printing/fcode.py
@@ -60,7 +60,7 @@ known_functions = {
     "conjugate": "conjg",
     "Max": "max",
     "Min": "min",
-    "Mod": "mod",
+    "Mod": "modulo",
 }
 
 

--- a/sympy/printing/tests/test_fcode.py
+++ b/sympy/printing/tests/test_fcode.py
@@ -1,6 +1,8 @@
 from sympy import (sin, cos, atan2, log, exp, gamma, conjugate, sqrt,
-    factorial, Integral, Piecewise, Add, diff, symbols, S, Float, Dummy, Eq,
-    Range, Catalan, EulerGamma, E, GoldenRatio, I, pi, Function, Rational, Integer, Lambda, sign)
+                   factorial, Integral, Piecewise, Add, diff, symbols, S,
+                   Float, Dummy, Eq, Range, Catalan, EulerGamma, E,
+                   GoldenRatio, I, pi, Function, Rational, Integer, Lambda,
+                   sign, Mod)
 
 from sympy.codegen import For, Assignment, aug_assign
 from sympy.codegen.ast import Declaration, Type, Variable, float32, float64, value_const, real, bool_, While
@@ -74,6 +76,8 @@ def test_fcode_Float():
 def test_fcode_functions():
     x, y = symbols('x,y')
     assert fcode(sin(x) ** cos(y)) == "      sin(x)**cos(y)"
+    assert fcode(Mod(x, y)) == "      mod(x, y)"
+    assert fcode(x % y) == "      mod(x, y)"
 
 
 def test_case():

--- a/sympy/printing/tests/test_fcode.py
+++ b/sympy/printing/tests/test_fcode.py
@@ -76,8 +76,8 @@ def test_fcode_Float():
 def test_fcode_functions():
     x, y = symbols('x,y')
     assert fcode(sin(x) ** cos(y)) == "      sin(x)**cos(y)"
-    assert fcode(Mod(x, y)) == "      mod(x, y)"
-    assert fcode(x % y) == "      mod(x, y)"
+    assert fcode(Mod(x, y)) == "      modulo(x, y)"
+    assert fcode(x % y) == "      modulo(x, y)"
 
 
 def test_case():

--- a/sympy/printing/tests/test_fcode.py
+++ b/sympy/printing/tests/test_fcode.py
@@ -76,10 +76,10 @@ def test_fcode_Float():
 def test_fcode_functions():
     x, y = symbols('x,y')
     assert fcode(sin(x) ** cos(y)) == "      sin(x)**cos(y)"
-    raises(ValueError, lambda: fcode(Mod(x, y), standard=66))
-    raises(ValueError, lambda: fcode(x % y, standard=66))
-    raises(ValueError, lambda: fcode(Mod(x, y), standard=77))
-    raises(ValueError, lambda: fcode(x % y, standard=77))
+    raises(NotImplementedError, lambda: fcode(Mod(x, y), standard=66))
+    raises(NotImplementedError, lambda: fcode(x % y, standard=66))
+    raises(NotImplementedError, lambda: fcode(Mod(x, y), standard=77))
+    raises(NotImplementedError, lambda: fcode(x % y, standard=77))
     for standard in [90, 95, 2003, 2008]:
         assert fcode(Mod(x, y), standard=standard) == "      modulo(x, y)"
         assert fcode(x % y, standard=standard) == "      modulo(x, y)"

--- a/sympy/printing/tests/test_fcode.py
+++ b/sympy/printing/tests/test_fcode.py
@@ -76,8 +76,13 @@ def test_fcode_Float():
 def test_fcode_functions():
     x, y = symbols('x,y')
     assert fcode(sin(x) ** cos(y)) == "      sin(x)**cos(y)"
-    assert fcode(Mod(x, y)) == "      modulo(x, y)"
-    assert fcode(x % y) == "      modulo(x, y)"
+    raises(ValueError, lambda: fcode(Mod(x, y), standard=66))
+    raises(ValueError, lambda: fcode(x % y, standard=66))
+    raises(ValueError, lambda: fcode(Mod(x, y), standard=77))
+    raises(ValueError, lambda: fcode(x % y, standard=77))
+    for standard in [90, 95, 2003, 2008]:
+        assert fcode(Mod(x, y), standard=standard) == "      modulo(x, y)"
+        assert fcode(x % y, standard=standard) == "      modulo(x, y)"
 
 
 def test_case():

--- a/sympy/utilities/codegen.py
+++ b/sympy/utilities/codegen.py
@@ -1204,7 +1204,7 @@ class FCodeGen(CodeGen):
                 assign_to = result.result_var
 
             constants, not_fortran, f_expr = self._printer_method_with_settings(
-                'doprint', dict(human=False, source_format='free'),
+                'doprint', dict(human=False, source_format='free', standard=95),
                 result.expr, assign_to=assign_to)
 
             for obj, v in sorted(constants, key=str):

